### DISCUSSION TITLE
Livewire enabled TinyMCE blade component.

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#2321] [2. Show and tell]
As TinyMCE was mentioned [here](https://github.com/livewire/livewire/issues/2285), [here](https://github.com/livewire/livewire/discussions/2273) and [here](https://github.com/livewire/livewire/discussions/2015) I thought I'd make a Livewire enabled TinyMCE blade component. The most up to date version can be found in this [gist](https://gist.github.com/ju5t/579a90b03f5cb80f3c6984d48a3385c4).

# Instructions

This is a very basic TinyMCE component. It uses 'entangle'. This allows you to link a Livewire and Alpine property to eachother. If one value changes, the other does too.

## Installation

Add `tinymce.blade.php` to `views/components/input`. This can be another component folder too if you prefer, but keep in mind that you should also change the `x-input.tinymce` structure accordingly.

Copy the following code in it:

```blade
<div
    x-data="{ value: @entangle($attributes->wire('model')) }"
    x-init="
        tinymce.init({
            target: $refs.tinymce,
            themes: 'modern',
            height: 200,
            menubar: false,
            plugins: [
                'advlist autolink lists link image charmap print preview anchor',
                'searchreplace visualblocks code fullscreen',
                'insertdatetime media table paste code help wordcount'
            ],
            toolbar: 'undo redo | formatselect | ' +
                'bold italic backcolor | alignleft aligncenter ' +
                'alignright alignjustify | bullist numlist outdent indent | ' +
                'removeformat | help',
            setup: function(editor) {
                editor.on('blur', function(e) {
                    value = editor.getContent()
                })

                editor.on('init', function (e) {
                    if (value != null) {
                        editor.setContent(value)
                    }
                })

                function putCursorToEnd() {
                    editor.selection.select(editor.getBody(), true);
                    editor.selection.collapse(false);
                }

                $watch('value', function (newValue) {
                    if (newValue !== editor.getContent()) {
                        editor.resetContent(newValue || '');
                        putCursorToEnd();
                    }
                });
            }
        })
    "
    wire:ignore
>
    <div>
        <input
            x-ref="tinymce"
            type="textarea"
            {{ $attributes->whereDoesntStartWith('wire:model') }}
        >
    </div>
</div>
```

## Usage

Make sure you have added the TinyMCE library in the layout that you're extending. Although with some extra work the component should be able to do this for you, it doesn't right now.

After the TinyMCE library has been added, this is how you use it:

```blade
<x-input.tinymce wire:model="editor" placeholder="Type anything you want..." />
```

Note that all HTML attributes will propogate to the text area. In this case the placeholder will automatically be passed on to the input in the blade component.

In the example we link it to `editor`. This can be anything really. But to demonstrate it's power, feel free to add `{{ $editor }}` below the blade component. As soon as you leave the editor (`blur`) the linked value will be updated for you and `{{ $editor }}` will change to.

**Comment by Whallas** [comment]
To create a two way data biding you can put this code in setup:
```js
function putCursorToEnd() {
    editor.selection.select(editor.getBody(), true);
    editor.selection.collapse(false);
}
$watch('value', function (newValue) {
    if (newValue !== editor.getContent()) {
        editor.resetContent(newValue || '');
        putCursorToEnd();
    }
});
```
_putCursorToEnd_ function will fix cursor position when _$watch_ sets the content.

**Reply by ju5t** [reply]
Oh that's cool!

**Comment by its-monotype** [comment]
```
editor.on('init', function (e) {
  if (value != null) {
    editor.setContent(value)
  }
})
```

to prevent this error
![image](https://user-images.githubusercontent.com/79363260/108595013-20eac200-7386-11eb-8b8b-f660b573f0c9.png)

**Reply by ju5t** [reply]
👍 

I have also updated the gist.

**Comment by faydanube** [comment]
![image](https://user-images.githubusercontent.com/8817046/109802276-8ed29b80-7c5a-11eb-89a7-fa0743dc128f.png)
`resources\views\welcome.blade.php`
```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <link rel="stylesheet" href="{{ asset('css/app.css') }}">
    <script src="{{ asset('js/tinymce_5.6.2_dev/tinymce.min.js') }}"></script>
    @livewireStyles
</head>
<body>
    @livewire('abc')

    <script src="{{ asset('js/app.js') }}"></script>
    @livewireScripts
</body>
</html>
```

`resources\views\components\input\tinymce.blade.php`
Same as the author

`resources\views\livewire\abc.blade.php`
```php
<div>
    <x-input.tinymce wire:model="editor" placeholder="Type anything you want..." />
</div>

```

I don't know why the error ...... can be replaced with the following
`resources\views\components\input\tinymce.blade.php`
```
<div>
    <script>tinymce.init({selector:'#content'});</script>
    <textarea name="" id="content" cols="30" rows="10"></textarea>
</div>
```

**Reply by ju5t** [reply]
Sorry, but I don't understand what you're trying to say?

**Reply by faydanube** [reply]
Error reported, editor did not render

**Reply by ju5t** [reply]
Do you have Alpine bundled in your app.js? This is a requirement. If not, install it using the guide on https://github.com/alpinejs/alpine.

**Reply by faydanube** [reply]
yes. I have Alpine bundle.

**Reply by ju5t** [reply]
Can you put your code up on Laravel playground? It is too difficult to debug it like this.

**Reply by smolDev-ai** [reply]
Did this ever get figured out? I'm having a similar issue.

**Reply by joshhanley** [reply]
@smolDev-ai what versions of Alpine and Livewire are you using?

**Reply by smolDev-ai** [reply]
> @smolDev-ai what versions of Alpine and Livewire are you using?

Livewire 2 and alpine 3

**Reply by joshhanley** [reply]
@smolDev-ai can you be specific about what patch versions you are using?

**Reply by smolDev-ai** [reply]
Sorry for the delay, I managed to get it working. I moved the tinymce script below the head, but above the livewireScripts directive and that fixed it for some reason.

**Reply by joshhanley** [reply]
@smolDev-ai ah good work! 

Generally the order I have scripts in at the bottom of the body, is
Livewire
Other libraries (or app.js bundle)
Alpine (or Alpine in the head but deferred, has same effect).

**Comment by midblep** [comment]
Just wanna drop a quick casual thank you for saving my life. This specific case has been bugging me for weeks, and this solution has worked perfectly and made me better understand data binding with entanglement in Alpine.

**Reply by ju5t** [reply]
Happy to help!

**Reply by adddz** [reply]
Thanks so much. You literally saved me hours of work. You're amazing.

**Comment by miatraxkhalifa** [comment]
how to place that component in a loop? adding a key to model name just renders it. 

   x-tinyMCE wire:model.defer="body.{{$value}}"  placeholder="Type anything you want..." /

**Comment by christogonus** [comment]
Hi, 

Thanks a lot for the gist. You saved me hours haha.

But I am stuck as submitting the new information doesn;t seem that the value of "editor" changed.

https://prnt.sc/yY48IdbgSdnN

https://prnt.sc/6d00mT2p96Z-

Could this be because of my version of TinyMCE (TinyCloud) ?

**Comment by vovarpd** [comment]
The gist worked fine up to Livewire 3.3.5
It stopped working with Livewire 3.4+.
TinyMCE is rendered correctly, but the model content is not updated anymore.

Any suggestions?

Thank you.

**Comment by Jagadish056** [comment]
Hey @ju5t, thanks for starting this discussion! Here's a reusable TinyMCE integration for Livewire while keeping Alpine.js and CSP in mind. 

Check it out! 

_form.blade.php_
```
<div x-data="tinymceEditor(@entangle('editor'))" wire:ignore>
    <textarea :id="$id('tinymce')" x-ref="tinymce" placeholder="Write something..."></textarea>
</div>
```
If you encounter an error when calling the Laravel Blade component, use `$wire.entangle()` instead of `@entangle()`.

```
<x-input label="Content" type="textarea" placeholder="Write something..." x-ref="tinymce" :attr="['x-data' => 'tinymceEditor($wire.entangle(\'editor\'))', 'wire:ignore' => '']" />
```

_resources\js\app.js_
```
Alpine.data('tinymceEditor', (entangle) => ({
    value: entangle,
    editor:null,
    init() {
        tinymce.init({

            // ...TinyMCE configuration

            target: this.$refs.tinymce,
            setup: (editor) => {

                editor.on('blur', (e) => {
                    this.value = editor.getContent()
                })

                editor.on('init', (e) => {
                    if (this.value != null) {
                        editor.setContent(this.value)
                    }
                })

                this.editor = editor
            }
        })

        // Or, this.editor = tinymce.get(this.$refs.tinymce.id)

        if (this.editor ) {
            this.$watch('value', (newValue) => {
                if (newValue !== this.editor.getContent()) {
                    this.editor.resetContent(newValue || '')
                    this.putCursorToEnd()
                }
            })
        }
    },

    putCursorToEnd() {
        this.editor.selection.select(this.editor.getBody(), true)
        this.editor.selection.collapse(false)
    }
}))
```
